banzai_nres/classify.py

```diff
@@ -55,23 +56,87 @@ def find_object_in_catalog(image, db_address, gaia_class, simbad_class):
+            results = restrict_simbad_results_to_stellar_only(results)
+            results = parse_simbad_coordinates(results)
+            results = convert_simbad_coordinates_to_degrees(results)
+            # note that results will be None if coordinate conversion fails
         except astroquery.exceptions.TableParseError:
             response = simbad_connection.last_response.content
             logger.error(f'Error querying SIMBAD. Response from SIMBAD: {response}', image=image)
             results = None
+
         if results:
-            image.classification = dbs.get_closest_phoenix_models(db_address, results[0]['Fe_H_Teff'],
-                                                                  results[0]['Fe_H_log_g'])[0]
+            results = get_closest_source(results, coordinate)
+            image.classification = dbs.get_closest_phoenix_models(db_address, results['Fe_H_Teff'],
+                                                                  results['Fe_H_log_g'])[0]
+            image.pm_ra = results['PMRA']  # note that we always assume these are in mas/yr... which they should be.
+            image.pm_dec = results['PMDEC']
             # Update the ra and dec to the catalog coordinates as those are basically always better than a user enters
             # manually.
-            image.ra, image.dec = results[0]['RA'], results[0]['DEC']
-            image.pm_ra, image.pm_dec = results[0]['PMRA'], results[0]['PMDEC']
+            image.ra = results['RA']
+            image.dec = results['DEC']
         # If there are still no results, then do nothing
 
 
+def get_closest_source(results, coordinate):
+    ra_unit, dec_unit = results['RA'].unit, results['DEC'].unit
+    coords = [SkyCoord(row['RA'], row['DEC'], unit=(ra_unit, dec_unit)) for row in results]
+    diffs = [coord.separation(coordinate).to(units.arcsecond).value for coord in coords]
+    closest = np.argmin(np.abs(diffs))
+    return results[closest]
+
+
+def restrict_simbad_results_to_stellar_only(results):
+    # remove galaxies
+    results = results[['*' in i for i in results['MAIN_ID']]]
+    is_nonstellar = np.zeros(len(results), dtype=bool)
+    nonstellar_otypes = ['su*', 'Pl?', 'Pl', 'SN', 'Galaxy', 'Planet', 'Planet?']
+    for i, row in enumerate(results):
+        if row['OTYPE'] in nonstellar_otypes:
+            is_nonstellar[i] = True
+
+    results = results[~is_nonstellar]
+    if len(results) == 0:
+        # in the (unlikely) event of no stellar objects being in the table, return None
+        # so that the later catch realizes this.
+        results = None
+    return results
+
+
+def parse_simbad_coordinates(results):
+    ra_unit, dec_unit = results['RA'].unit, results['DEC'].unit
+    units_not_recognized = type(ra_unit) is units.UnrecognizedUnit or type(dec_unit) is units.UnrecognizedUnit
+    if units_not_recognized:
+        # if the units are unrecognized (see if they are something descriptive like hms or dms.
```

Comment:
Very minor comment, but there is a missing ) in the comment.

Reply:
Thanks! Fixed.